File: dbxml/examples/java/basic/QueryTxn.java
/*
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 2004,2009 Oracle.  All rights reserved.
 *
 */

package basic;
import java.io.File;

import com.sleepycat.dbxml.XmlException;
import com.sleepycat.dbxml.XmlManager;
import com.sleepycat.dbxml.XmlManagerConfig;
import com.sleepycat.dbxml.XmlContainer;
import com.sleepycat.dbxml.XmlContainerConfig;
import com.sleepycat.dbxml.XmlDocument;
import com.sleepycat.dbxml.XmlUpdateContext;
import com.sleepycat.dbxml.XmlQueryExpression;
import com.sleepycat.dbxml.XmlResults;
import com.sleepycat.dbxml.XmlValue;
import com.sleepycat.dbxml.XmlQueryContext;

import com.sleepycat.dbxml.XmlTransaction;
import com.sleepycat.db.Environment;
import com.sleepycat.db.EnvironmentConfig;
import com.sleepycat.db.DatabaseException;

/**
 * Query is a very simple Berkeley DB XML program
 * that performs a query and handles results.
 * It demonstrates transactional initialization and container creation,
 * transactional document insertion, transactional query creation 
 * and execution, use of a variable in a query and context, and
 * transactional results handling
 * <p>
 * To run the example:
 * <pre>
 * java basic.QueryTxn [-h environmentDirectory]
 * </pre>
 */
class queryTxn
{
    static void usage()
    {
	System.err.println("Usage: java dbxml.basic.queryTxn [-h environmentDirectory]");
	System.exit(-1);
    }

    // This function is used to ensure that databases are
    // properly closed, even on exceptions
    private static void cleanup(Environment env,
				XmlManager mgr, 
				XmlContainer cont) {
	try {
	    if (cont != null)
		cont.delete();
	    if (mgr != null)
		mgr.delete();
	    if (env != null)
		env.close();
	} catch (Exception e) {
	    // ignore exceptions in cleanup
	}
    }

    private static Environment createEnvironment(String home)
	throws Throwable {
	EnvironmentConfig config = new EnvironmentConfig();
	config.setTransactional(true);
	config.setAllowCreate(true);
	config.setRunRecovery(true);
	config.setInitializeCache(true);
	config.setCacheSize(25 * 1024 * 1024); // 25MB cache
	config.setInitializeLocking(true);
	config.setInitializeLogging(true);
	config.setErrorStream(System.err);
	File f = new File(home);
	return new Environment(f, config);
    }

    public static void main(String args[])
	throws Throwable {

	// This program uses a named container, which will apear
	// on disk
	String containerName = "people.dbxml";
	String content = "<people><person><name>joe</name></person><person><name>mary</name></person></people>";
	String docName = "people";
	// Note that the query uses a variable, which must be set
	// in the query context
	String queryString =
	    "collection('people.dbxml')/people/person[name=$name]";
	String environmentDir = ".";

	if (args.length == 2) {
	    environmentDir = args[1];
	} else if (args.length != 0)
	    usage();

	XmlManager mgr = null;
	XmlContainer cont = null;
	Environment env = null;
	try {
	    // Create and open a Berkeley DB Transactional Environment.
	    env = createEnvironment(environmentDir);

	    // All BDB XML programs require an XmlManager instance.
	    // Create it from the DB Environment, but do not adopt the
	    // Environment
	    XmlManagerConfig mconfig = new XmlManagerConfig();
	    mconfig.setAllowExternalAccess(true);
	    mgr = new XmlManager(env, mconfig);

	    
	    // Because the container will exist on disk, remove it
	    // first if it exists
	    if (mgr.existsContainer(containerName) != 0)
		mgr.removeContainer(containerName);
		
	    // Create a container that is transactional.  Specify
	    // that it is also a Node Storage container, with nodes
	    // indexed
	    XmlContainerConfig cconfig = new XmlContainerConfig();
	    cconfig.setContainerType(XmlContainer.NodeContainer);
	    cconfig.setIndexNodes(XmlContainerConfig.On);
	    cconfig.setTransactional(true);
	    cont = mgr.createContainer(containerName, cconfig);
	    
	    // Perform the putDocument in a transaction, created
	    // from the XmlManager
	    XmlTransaction txn = mgr.createTransaction();
	    
	    cont.putDocument(txn, docName, content);
	    
	    // commit the Transaction
	    txn.commit();
	    
	    // Querying requires an XmlQueryContext
	    XmlQueryContext qc = mgr.createQueryContext();
	    
	    // Add a variable to the query context, used by the query
	    qc.setVariableValue("name", new XmlValue("mary"));
	    
	    // Create a new transaction for the query
	    txn = mgr.createTransaction();
	    
	    // Note the passing of txn to both methods
	    XmlQueryExpression expr = mgr.prepare(txn, queryString, qc);
	    XmlResults res = expr.execute(txn, qc);
	    
	    // Note use of XmlQueryExpression::getQuery() and
	    // XmlResults::size()
	    System.out.println("The query, '" + expr.getQuery() +
			       "'\n\t returned " + res.size() + " result(s)");
	    
	    // Process results -- just print them
	    XmlValue value = new XmlValue();
	    System.out.print("Result: ");
	    while ((value = res.next()) != null) {
		System.out.println("\t" + value.asString());
	    }
	    
	    // done with the transaction
	    txn.commit();
	    
	    // explicitly delete objects to release resources
	    res.delete();
	    expr.delete();
	} 
	catch (XmlException xe) {
	    System.err.println("XmlException during helloWorldTxn: " + 
			       xe.getMessage());
	    throw xe;
	}
	catch (DatabaseException de) {
	    System.err.println("DatabaseException during helloWorldTxn: " + 
			       de.getMessage());
	    throw de;
	}
	finally {
	    cleanup(env, mgr, cont);
	}
    }
}

File: dbxml/examples/python/basic/queryTxn.py
"""
 * Query is a very simple Berkeley DB XML program
 * that performs a query and handles results transactionally.
 * It demonstrates transactional initialization container creation,
 * transactional document insertion, query creation and execution,
 * use of a variable in a query and context, and
 * transactional results handling.
 *
 * To run the example:
 *     python queryTxn.py
"""
from dbxml import *
from bsddb3.db import *

def createEnvironment(home):
    """Create DBEnv and initialize XmlManager"""
    try:
        environment = DBEnv()
        # initialize DBEnv for transactions
        environment.open(home, DB_RECOVER|DB_CREATE|DB_INIT_LOCK|
                         DB_INIT_LOG|DB_INIT_MPOOL|DB_INIT_TXN, 0)
    except DBError, exc:
        print exc
        sys.exit()
    try:
        mgr = XmlManager(environment, 0)
    except XmlException, xe:
        print xe
        sys.exit()
    return mgr

def createContainer(mgr, containerName, flags):
    """create/open a node container"""
    if mgr.existsContainer(containerName) != 0:
        mgr.removeContainer(containerName)
    try:
        return mgr.openContainer(containerName,
                                 flags|DB_CREATE,
                                 XmlContainer.NodeContainer)
    except XmlException, ex:
        print ex
        sys.exit()

def queryTxn(mgr, cont):
    content = "<people><person><name>joe</name></person><person><name>mary</name></person></people>"
    docName = "people"
    queryString = "collection('%s')/people/person[name=$name]"%cont.getName()
    try:
        # all modification operations need XmlUpdateContext
        uc = mgr.createUpdateContext()

        txn = mgr.createTransaction()
        try:
            docName = cont.putDocument(txn, docName, content, uc, DBXML_GEN_NAME)
            txn.commit()
        except XmlException, ex:
            print ex
            txn.abort()

        # querying requires an XmlQueryContext
        qc = mgr.createQueryContext()

        # add a variable to the context, as required by the query
        qc.setVariableValue("name", XmlValue("mary"))

        # create a new transaction for the query
        txn = mgr.createTransaction()

        try:
            # note: these two calls could be replaced by a shortcut
            # mgr.query(txn, queryString, qc), but under the covers,
            # this is what is happening.  Prepared queries can be shared
            # and reused
            expr = mgr.prepare(txn, queryString, qc)
            res = expr.execute(txn, qc)
	    
            # note use of XmlQueryExpression.getQuery() and
            # XmlResults.size()
            print "The query '%s'\n\t returned %d result(s)"%(expr.getQuery(),res.size())
	    
            # process results -- just print them.  XmlResults
            # are a Python iterator
            for value in res:
                print "Result: %s"%value.asString()
            txn.commit()
        except XmlException, xe:
            print xe
            txn.abort()
    except XmlException, inst:
        print inst

def usage():
    print "queryTxn.py [-h home_directory]"

# "main"
import sys
import getopt
def main():
    home = "."
    # parse command line options
    try:
        opts, args = getopt.getopt(sys.argv[1:], "h:")
    except getopt.error, msg:
        print msg
        usage()
        sys.exit(2)
    for o, a in opts:
        if o == "-h":
            home = a

    # some configuration...
    containerName = "people.dbxml"
    # initialize...
    mgr = createEnvironment(home)
    # create/open a transactional container
    cont = createContainer(mgr, containerName, 
                           DBXML_TRANSACTIONAL)

    # run the guts of the program...
    queryTxn(mgr, cont)

if __name__ == "__main__":
    main()
